### PR TITLE
Fix/qwen2.5 vl skeleton

### PR DESCRIPTION
**Pull request by AdityaShome**
### Problem Solved
The kornia Qwen2.5-VL architecture had critical mismatches with HuggingFace:
- Only 50% of pretrained weights loaded (198/394 parameters)
- PyTorch-ONNX numerical differences were massive (~40-46, tolerance is 0.0001)

### Root Cause
The MLP in transformer blocks used simple 2-layer architecture while HuggingFace uses gated SwiGLU.

## 🛠️ Changes Made
- [x] Replaced simple 2-layer MLP with gated SwiGLU architecture (gate_proj, up_proj, down_proj)
- [x] Fixed MLP intermediate_size from 5120 to 3420 to match HuggingFace
- [x] Removed bias from all LayerNorm layers
- [x] Removed LayerNorm from PatchMerger
- [x] Removed bias from PatchMerger Conv3d
- [x] Updated weight loader key mappings for gated MLP
- [x] Updated tests to match new architecture

---

## 🧪 How Was This Tested?
- [x] **Unit Tests:** 
  - Updated [test_qwen2_vl.py](cci:7://file:///d:/gsoc%202/kornia-whole/kornia/tests/models/qwen25/test_qwen2_vl.py:0:0-0:0) and [test_qwen25_vision.py](cci:7://file:///d:/gsoc%202/kornia-whole/kornia/tests/models/qwen25/test_qwen25_vision.py:0:0-0:0) 
  - All tests pass: model creation, forward pass, batch processing (1,2,4), multiple resolutions (224x224, 448x448)
- [x] **Manual Verification:** 
  - Ran [validate_pytorch_onnx.py](cci:7://file:///d:/gsoc%202/kornia-whole/kornia/validate_pytorch_onnx.py:0:0-0:0) - PyTorch vs ONNX numerical accuracy within tolerance (0.000086 < 0.0001)
  - Verified weight loading: 390/390 parameters (100%, was 198/394)
- [x] **Performance/Edge Cases:** 
  - Tested with batch sizes 1, 2, 4
  - Tested with resolutions 224x224 and 448x448
  - Validated output shapes match expected (B, seq_len, 2048)

---

## 🕵️ AI Usage Disclosure
*Check one of the following:*
- [ ] 🟢 **No AI used.**
- [x] 🟡 **AI-assisted:** I used AI for boilerplate/refactoring but have manually reviewed and tested every line.
- [ ] 🔴 **AI-generated:** (Note: These PRs may be subject to stricter scrutiny or immediate closure if the logic is not explained).

---

## 🚦 Checklist
- [ ] I am assigned to the linked issue (required before PR submission)
- [ ] The linked issue has been approved by a maintainer
- [x] I have performed a **self-review** of my code (no "ghost" variables or hallucinations).
- [x] My code follows the existing style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] (Optional) I have attached screenshots/recordings for UI changes.

---

### Results After Fix
- Weight Loading: 100% (390/390 parameters)
- Numerical Accuracy: Max diff 0.000086 (within tolerance)
- All Tests Pass: Forward pass, batching, multiple resolutions

### CODE REVIEW

File: kornia/models/qwen25/qwen2_vl.py [hunk 1]
```diff
@@ -26,24 +26,152 @@
...
+        context_window: int = 224,
+        spatial_merge_size: int = 2,
+        in_channels: int = 3,
+    ) -> None:
         super().__init__()
-        self.conv = nn.Conv2d(3, dim, kernel_size=14, stride=14)
-        self.ln_q = nn.LayerNorm(dim, eps=1e-6)
+        self.spatial_merge_size = spatial_merge_size
+
+        # Use Conv3d to match HuggingFace (handles temporal dimension)
```
Comment (edgarriba) [inline kornia/models/qwen25/qwen2_vl.py]: please keep comments agnostic to refe other libraries

File: kornia/models/qwen25/qwen2_vl.py [hunk 2]
```diff
@@ -82,6 +215,10 @@ class Qwen2VLVisionAttention(Module):
 
     def __init__(self, dim: int, num_heads: int = 16) -> None:
         super().__init__()
+
+        # Validate that dimension is divisible by number of heads
+        assert dim % num_heads == 0, f"dim ({dim}) must be divisible by num_heads ({num_heads})"
```
Comment (edgarriba) [inline kornia/models/qwen25/qwen2_vl.py]: can we use kornia check module

File: kornia/models/qwen25/weights_loader.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+"""Weight loader for converting HuggingFace Qwen2.5-VL weights to kornia format."""
+
+from __future__ import annotations
+
+import re
+
+import torch
+
+
+class Qwen25WeightLoader:
```
Comment (edgarriba) [inline kornia/models/qwen25/weights_loader.py:28]: cool -- I can see this being a good candidate to go under kornia.models directly to be reused by other models (in a future PR)

File: kornia/models/qwen25/qwen2_vl.py [hunk 4]
```diff
@@ -19,56 +19,160 @@
...
+        # Spatial compression: group 2x2 patches
+        # Input: (B, H*W, C) where L=H*W is a perfect square
+        B, L, C = x.shape
+
+        # Calculate grid size from sequence length
+        # Warning: This uses Python int() which gets traced as a constant in ONNX
+        # For ONNX export, the model is traced with a specific resolution
+        # and that resolution is baked into the graph. Dynamic resolution
+        # support would require exporting multiple ONNX models or using torch.jit.script
+        import math
```
Comment (edgarriba) [inline kornia/models/qwen25/qwen2_vl.py]: imports go on the top ....

**Comment by AdityaShome** [comment]
Vision encoder is implemented by splitting it into 34 separate components due to a critical PyTorch bug where exporter fails to properly trace parameters that were loaded via load_state_dict(). To work around this, each submodule was exported individually (patch_embed, 32 transformer blocks, and merger) directly from the pretrained model, which preserves weights correctly.

The merger component  for dynamic aspect ratios, created a wrapper that accepts grid_h and grid_w as 0 dimensional tensor inputs rather than Python integers, allowing tracer to create dynamic shape operations in the graph instead of hardcoding dimensions as constants. It also projects the features into text embedding space for which I will build the decoder, when it is approved.
During inference, we pass grid dimensions directly to the onnx merger, enabling exported onnx to handle any aspect ratio (224×224, 336×224, 224×336, etc.) which is tested with numerical equivalence to pytorch (validated at max_diff=0.000086).
